src/gflownet/tasks/qm9/qm9.py

```diff
@@ -139,86 +99,48 @@ def set_default_hps(self, cfg: Config):
         cfg.opt.lr_decay = 20000
         cfg.opt.clip_grad_type = "norm"
         cfg.opt.clip_grad_param = 10
+        cfg.algo.max_nodes = 9
         cfg.algo.global_batch_size = 64
         cfg.algo.train_random_action_prob = 0.001
         cfg.algo.illegal_action_logreward = -75
         cfg.algo.sampling_tau = 0.0
         cfg.model.num_emb = 128
         cfg.model.num_layers = 4
-        cfg.task.qm9.temperature_sample_dist = "uniform"
-        cfg.task.qm9.temperature_dist_params = [0.5, 32.0]
-        cfg.task.qm9.num_thermometer_dim = 32
-
-    def setup(self):
-        RDLogger.DisableLog("rdApp.*")
-        self.rng = np.random.default_rng(142857)
-        self.env = GraphBuildingEnv()
-        self.ctx = MolBuildingEnvContext(["H", "C", "N", "F", "O"], num_cond_dim=32)
-        self.training_data = QM9Dataset(self.cfg.task.qm9.h5_path, train=True, target="gap")
-        self.test_data = QM9Dataset(self.cfg.task.qm9.h5_path, train=False, target="gap")
+        cfg.cond.temperature.sample_dist = "uniform"
+        cfg.cond.temperature.dist_params = [0.5, 32.0]
+        cfg.cond.temperature.num_thermometer_dim = 32
 
-        model = GraphTransformerGFN(self.ctx, self.cfg)
-        self.model = model
-        # Separate Z parameters from non-Z to allow for LR decay on the former
-        Z_params = list(model.logZ.parameters())
-        non_Z_params = [i for i in self.model.parameters() if all(id(i) != id(j) for j in Z_params)]
-        self.opt = torch.optim.Adam(
-            non_Z_params,
-            self.cfg.opt.learning_rate,
-            (self.cfg.opt.momentum, 0.999),
-            weight_decay=self.cfg.opt.weight_decay,
-            eps=self.cfg.opt.adam_eps,
-        )
-        self.opt_Z = torch.optim.Adam(Z_params, self.cfg.opt.learning_rate, (0.9, 0.999))
-        self.lr_sched = torch.optim.lr_scheduler.LambdaLR(self.opt, lambda steps: 2 ** (-steps / self.cfg.opt.lr_decay))
-        self.lr_sched_Z = torch.optim.lr_scheduler.LambdaLR(
-            self.opt_Z, lambda steps: 2 ** (-steps / self.cfg.opt.lr_decay)
+    def setup_env_context(self):
+        self.ctx = MolBuildingEnvContext(
+            ["H", "C", "N", "F", "O"], expl_H_range=[0, 1, 2, 3], num_cond_dim=32, allow_5_valence_nitrogen=True
```

Comment:
are hydrogen atoms explicitly added in the dataset already (or are 3D coordinates required in the task) ? If not, having hydrogen addition as explicit action `H` might not be required.

Reply:
Ah you're right the `H` action is not needed here.